homeassistant/components/zwave_js/update.py

```diff
@@ -0,0 +1,199 @@
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Z-Wave button from config entry."""
+    client: ZwaveClient = hass.data[DOMAIN][config_entry.entry_id][DATA_CLIENT]
+
+    @callback
+    def async_add_firmware_update_entity(node: ZwaveNode) -> None:
+        """Add firmware update entity."""
+        driver = client.driver
+        assert driver is not None  # Driver is ready before platforms are loaded.
+        async_add_entities([ZWaveNodeFirmwareUpdate(driver, node)], True)
+
+    config_entry.async_on_unload(
+        async_dispatcher_connect(
+            hass,
+            f"{DOMAIN}_{config_entry.entry_id}_add_firmware_update_entity",
+            async_add_firmware_update_entity,
+        )
+    )
+
+
+def get_latest_firmware(
+    firmwares: list[FirmwareUpdateInfo],
+) -> FirmwareUpdateInfo:
+    """Get the latest firmware from the available updates."""
+    latest_firmware = FirmwareUpdateInfo("0.0.0", "", [])
+    for firmware in firmwares:
+        if AwesomeVersion(firmware.version) > AwesomeVersion(latest_firmware.version):
+            latest_firmware = firmware
+    return latest_firmware
+
+
+class ZWaveNodeFirmwareUpdate(UpdateEntity):
+    """Representation of a firmware update entity."""
+
+    _attr_entity_category = EntityCategory.CONFIG
+    _attr_device_class = UpdateDeviceClass.FIRMWARE
+    _attr_supported_features = (
+        UpdateEntityFeature.INSTALL
+        | UpdateEntityFeature.RELEASE_NOTES
+        | UpdateEntityFeature.SPECIFIC_VERSION
+    )
+    _attr_has_entity_name = True
+
+    def __init__(self, driver: Driver, node: ZwaveNode) -> None:
+        """Initialize a Z-Wave device firmware update entity."""
+        self.driver = driver
+        self.node = node
+        self.available_firmware_updates: list[FirmwareUpdateInfo] = []
+        self._latest_version_firmware: FirmwareUpdateInfo | None = None
+
+        # Entity class attributes
+        self._attr_name = "Firmware"
+        self._base_unique_id = get_valueless_base_unique_id(driver, node)
+        self._attr_unique_id = f"{self._base_unique_id}.firmware_update"
+        # device may not be precreated in main handler yet
+        self._attr_device_info = DeviceInfo(
+            identifiers={get_device_id(driver, node)},
+            sw_version=node.firmware_version,
+            name=node.name or node.device_config.description or f"Node {node.node_id}",
+            model=node.device_config.label,
+            manufacturer=node.device_config.manufacturer,
+            suggested_area=node.location if node.location else None,
+        )
+
+        self._attr_installed_version = self._attr_latest_version = node.firmware_version
+
+    @property
+    def extra_state_attributes(self) -> Mapping[str, Any] | None:
+        """Return entity specific state attributes."""
+
+        return {
+            ATTR_AVAILABLE_FIRMWARE_UPDATES: [
```

Comment:
I don't think that we have to keep these around in the state machine.

Reply:
Part of our past discussion was that users need to be able to pick a version to install because of the way that manufacturers do firmware files. This is the only way to expose the available versions using the firmware update service and update entity that I can think of but open to suggestions

Reply:
Couldn't they browse the Z-Wave JS Firmware website for this?

Reply:
there is no website (yet). Maybe there will be one in the future? For now all they would have to go by is the repo and they'd have to manually figure out which versions are available for them based on their current firmware version

Reply:
Let's limit the update entity should be focused on just updating to the latest version. If we want to offer other available versions and updating those, add that as a feature to the Z-Wave JS panel.